### ISSUE TITLE
File loader only works when bundling

### ISSUE THREAD

**Issue by oliverjam** [evanw/esbuild#1528]
Is it expected behaviour that non-JS imports are not transformed if the importing file isn't bundled? E.g.

```js
// src/index.js
const img = require("./example.png");
```

Running esbuild with these options:
```sh
esbuild src/* --loader:.png=file --outdir=out
```

Results in this output:
```
out
├── example-XAKSQQ43.png
├── example.js
└── index.js
```
```js
// out/example.js
module.exports = "./example-XAKSQQ43.png";
```
```js
// out/index.js
const img = require("./example.png");
```

Since the path is unchanged Node will crash when attempting to import the png:
```
Error: Cannot find module './example.png'
```

I've tried messing around with plugins to manually transform the import path but it seems like I can only get at it before it is resolved.

---

I understand esbuild is primarily focused on the bundling use-case. However that's not always practical for server-side code (e.g. in my case the server relies on having a `pages/` directory at runtime).

If this worked it would be a really fast, easy way to use immutable hashed assets in Node. I ended up writing my own plugin that just hashes the files and copies them over manually, then outputs a `manifest.json` that the server has to read, which is much less elegant than using imports.

**Comment by evanw** [comment]
> Is it expected behaviour that non-JS imports are not transformed if the importing file isn't bundled?

Yes. By design, import paths are only inspected when bundling is active.

However, just because bundling is active doesn't necessarily mean import paths are bundled. Import paths that have been marked as "external" are not bundled. So you can enable bundling and then write an on-resolve plugin that intercepts import paths, marks them as `external: true` to avoid bundling them, and then do whatever you want with them (leave them alone, change them to another path, add arbitrary custom side effects).

**Comment by oliverjam** [comment]
In that case is it worth logging a warning if someone tries to use the file loader without bundling enabled?

I'm a bit confused by your suggestion sorry. If I enable bundling with a single entrypoint, then use a plugin to intercept `.png` resolves and mark them as external, esbuild no longer hashes and copies through the `.png` file. Were you suggesting to implement that part manually?

Thanks for your response by the way, esbuild has been a joy to work with (especially as a frontend-focused dev who literally never got their head around WebPack 😅)

**Comment by CyriacBr** [comment]
I encountered this issue as well and was very confused why the built-in JSON loader wasn't working.  It'd be indeed interesting to log a warning with such cases, and also for plugins that are only able to run when bundling.
